latex_engine_cmds.bzl

```diff
@@ -0,0 +1,258 @@
+        "--env",
+        files,
+        join_with = ":",
+        format_joined = "LUAINPUTS=%s:",
+        map_each = to_dirname,
+        uniquify = True,
+    )
+    common_args.add_joined(
+        "--env",
+        files,
+        join_with = ":",
+        format_joined = "TEXFONTMAPS=:%s:",
+        map_each = to_dirname,
+        uniquify = True,
+    )
+    common_args.add_joined(
+        "--env",
+        [
+            dep_tools[0].to_list()[0].dirname,  # latex bin folder
+            dep_tools[4].to_list()[0].dirname,  # script folder
+            "/bin",  # sed, rm, etc. needed by mktexlsr
+            "/usr/bin",  # needed to find pytohn
+        ],
+        join_with = ":",
+        format_joined = "PATH=%s:.:",
+        uniquify = True,
+    )
+    tex_env_dict = {
+        "SOURCE_DATE_EPOCH": ["0"],
+        "TEXMF": ["$PWD/external"],
+        "TEXMFCNF": [ctx.files.web2c[0].dirname],
+        "TEXMFHOME": ["$PWD"],
+        "TEXMFROOT": ["."],
+    }
+    for k, v in tex_env_dict.items():
+        common_args.add_joined(
+            "--env",
+            v,
+            join_with = ":",
+            format_joined = "{}=%s".format(k),
+        )
+
+    for dep_tool in dep_tools:
+        common_args.add_all("--dep-tool", dep_tool)
+    return common_args
+
+def add_flags(ctx):
+    """
+    Set flags for supplying arguments to latexrun and the engine.
+
+    Args:
+      ctx: For accessing the flags
+
+    Returns:
+      Flags for the latexrun command
+    """
+    prefix = "--flag="
+    cmd = [
+        "--latex-args=--progname={}".format(ctx.attr._progname),
+        "--output-format={}".format(ctx.attr.format),
+        "--shell-escape",
+        "--jobname={}".format(ctx.label.name),
+        "--fmt={}.fmt".format(ctx.attr._progname),
+    ]
+
+    flags = [
+        prefix + " ".join(cmd),
+    ]
+
+    for value in ctx.attr.cmd_flags:
+        if "output-format" in value and ctx.attr.format not in value:
+            fail("Value of attr format ({}) conflicts with value of flag {}".format(ctx.attr.format, value))
+        flags.append(prefix + value)
+    return flags
+
+def lualatex_engine_cmd_gen(ctx, dep_tools, latex_tool):
+    """
+    Generate commands specific for the lualatex engine.
+
+    This function generates three commands:
+    - generate format file
+    - create ls-R database
+    - call latex to build the document
+
+    Args:
+      ctx: For accessing the inputs parameters.
+      dep_tools: the tools needed by latexrun
+      latex_tool: The latex engine to use, in this case we make the
+        assumption that it is the luahbtex binary (This should be
+        enforced more strongly).
+
+    Returns:
+      A list with the three commands
+    """
+
+    # TODO: Each command could be its own invokation to ctx.actions.run
+    #       this way the artifacts could be cache, it might not help that
+    #       much though, as any change to the inputs is likely to re-trigger
+    #       a re-build.
```

Comment:
Regarding rebuilds: I'm not too familiar with the inner workings of LaTeX compilers, but if the first command generates an intermediate representation of sorts, it definitely would make sense to separate the different commands - if someone fixes e.g. indentation in a source file, any intermediate representation should likely be the same, hence avoiding future steps.

(Note that I haven't read the implementation yet...)

Reply:
Something along those lines.
No I'm not well versed in the inner workings of latex either. We should probably ask someone from the latex community that knows their way around the tex distro code to review this code. I feel like we're aiming in the dark.